pkg/printers/sarif.go

```diff
@@ -0,0 +1,212 @@
+// to generate a sarif report.
+//
+// JSON structure:
+//
+//		type data struct {
+//					Violations []violation `json:"violations`
+//		}
+//
+//		type violation struct {
+//		      Id          string `json:"id"`
+//	          Description string `json:"description"`
+//	          Reason      string `json:"reason"`
+//	          Severity    string `json:"severity"`
+//		}
+//
+// Example:
+//
+//	 expressions := map[string] {
+//			RuleJSONPathExpressionKey: "violations.#.id",
+//			DescriptionJSONPathExpressionKey: "violations.#.description",
+//		    HelpJSONPathExpressionKey: "violations.#.reason",
+//		    SeverityJSONPathExpressionKey: "violations.#.severity",
+//		}
+//
+// For an example sarif report, see testdata/sarif_report.json.
+//
+// Advanced usages:
+// For constructing multiline help messages with values from different JSON fields, see roxctl/image/check/check.go
+// as an example usage of constructing 1) a rule ID with multiple values using gjson.TextModifier and 2) creating the
+// multiline help text via gjson.TextModifier.
+func NewSarifPrinter(jsonPathExpressions map[string]string, entity string, reportType string) *SarifPrinter {
+	return &SarifPrinter{jsonPathExpressions: jsonPathExpressions, entity: entity, reportType: reportType}
+}
+
+// Print will create a sarif report from the given object and write the output to the given io.Writer.
+func (s *SarifPrinter) Print(object interface{}, out io.Writer) error {
+	sarifEntries, err := sarifEntriesFromJSONObject(object, s.jsonPathExpressions)
+	if err != nil {
+		return err
+	}
+
+	run := sarif.NewRunWithInformationURI("roxctl", "https://github.com/stackrox/stackrox")
+	// Set the version information and the full name on the tool driver.
+	run.Tool.Driver.
+		WithVersion(version.GetMainVersion()).
+		WithFullName("roxctl command line utility")
+
+	report, err := sarif.New(sarif.Version210)
+	if err != nil {
+		return errors.Wrap(err, "creating sarif report")
+	}
+
+	for _, entry := range sarifEntries {
+		addEntry(run, entry, s.entity, s.reportType)
+	}
+
+	report.AddRun(run)
+	return report.PrettyWrite(out)
+}
+
+func addEntry(run *sarif.Run, entry sarifEntry, entity string, name string) {
+	run.AddRule(entry.ruleID).
+		WithName(name).
+		WithShortDescription(sarif.NewMultiformatMessageString(entry.description)).
+		WithFullDescription(sarif.NewMultiformatMessageString(entry.description)).
+		WithHelp(sarif.NewMultiformatMessageString(entry.help))
+
+	run.AddResult(sarif.NewRuleResult(entry.ruleID).
+		WithLevel(toSarifLevel(entry.severity)).
+		// Reusing the help here, since the help includes remediation information.
+		WithMessage(sarif.NewMessage().WithText(entry.help)).
+		WithLocations([]*sarif.Location{
+			{
+				PhysicalLocation: &sarif.PhysicalLocation{
+					ArtifactLocation: sarif.NewArtifactLocation().WithUri(entity),
+					Region: sarif.NewRegion().
```

Comment:
`NewSimpleRegion(startLine, endLine int)` ?
Do we need to specify all of it? I know it's required but even github examples does not contain end linte. https://docs.github.com/en/code-security/code-scanning/integrating-with-code-scanning/sarif-support-for-code-scanning#physicallocation-object

Reply:
Works as well, not sure why GH specifies it as required 🤷